### DISCUSSION TITLE
Two input sources into single output

### DISCUSSION THREAD

**Discussion by Mukhin22** [pschatzmann/arduino-audio-tools#721] [Q&A] [answered]
Hey, Phil, is it possible to combine simple microphone input with A2DP stream input and route it into the single output i2s speaker with amplifier using your library. I'm using ESP32-DevKitC(controller), MAX98357A(amplifier),  INMP441(micro). And I've already managed to get 1)i2s in to i2s out and 2) a2dp in to i2s out working separately. And now I want to combine it. Just asking for your advice.

**Comment by pschatzmann** [comment]
I wouldn't see any reason why this should not work. 
Be careful with the INMP441 because you will need to do some extra steps to get a clean 16bit signal.
I would also suggest that you use the basic A2DP examples (which use a callback) instead of the stream implementation which has quite some overhead...

ps. I assume you plan to use 2 different ESP32 to implement this...

**Reply by Mukhin22** [reply]
No, I'm planning to use single ESP32 and maybe you have some suggestion how to implement such configuration. I2S[IN] microphone + AD2P[IN]  sound stream  -> I2S[OUT] amplifier with speaker out. So it's like 2 inputs into single output.

**Reply by pschatzmann** [reply]
Ok, then I suggest that you try to use the InputMixer

**Reply by Mukhin22** [reply]
Oh, thank you, found it in examples of audiokit. I'll try.

**Reply by Mukhin22** [reply]
In basic examples, that are more preffered to be used, I can see that a2dp input used with BluetoothA2DPSink class, not with A2DPStream.
But when I wan't to use InputMixer API I need to pass Stream object into the mixer.add() method.
How can I get A2DP stream object from BluetoothA2DPSink to connect it to the InputMixer? Because i can't just use like that mixer.add(a2dp_sink_input_obj)

**Reply by pschatzmann** [reply]
You would need to use the A2DPStream class. 
There are plenty of example that show how to use this class.
If you want to use the callback function you would need to mix with the OutputMixer
see https://github.com/pschatzmann/arduino-audio-tools/wiki/Splitting-and-Merging-Audio

**Comment by Mukhin22** [comment]
I'm trying to do something like that:

```
#define USE_A2DP_STREAM 1

AudioBaseInfo info;
const uint16_t sample_rate = 44100;
const uint16_t channels = 2;
const uint16_t bits_per_sample = 16;
I2SStream i2s_input;
I2SStream i2s_output;
InputMixer<int16_t> mixer;

#if USE_A2DP_STREAM
A2DPStream a2dp_input;
#else 
BluetoothA2DPSink a2dp_in_sink;
// Write data to SPDIF in callback
void read_data_stream(const uint8_t *data, uint32_t length) {
  i2s_output.write(data, length);
}
#endif /* USE_A2DP_STREAM */

StreamCopy copier(mixer, i2s_output);

void setup(void) {  
  auto i2s_input_config = i2s_input.defaultConfig(RX_MODE);
  i2s_input_config.sample_rate = sample_rate; 
  i2s_input_config.bits_per_sample = bits_per_sample; 
  i2s_input_config.i2s_format = I2S_STD_FORMAT;
  i2s_input_config.is_master = true;
  i2s_input_config.port_no = 0;
  i2s_input_config.pin_bck = 15;
  i2s_input_config.pin_ws = 14;
  i2s_input_config.pin_data = 22;
  i2s_input.begin(i2s_input_config);

  auto i2s_output_config = i2s_output.defaultConfig(TX_MODE);
  i2s_output_config.copyFrom(info);
  i2s_output_config.i2s_format = I2S_STD_FORMAT;
  i2s_output_config.is_master = true;
  i2s_output_config.port_no = 1;
  i2s_output_config.pin_bck = 18;
  i2s_output_config.pin_ws = 19;
  i2s_output_config.pin_data = 23;
  i2s_output.begin(i2s_output_config);

#if USE_A2DP_STREAM
  auto a2dp_cfg = a2dp_input.defaultConfig(RX_MODE);
  a2dp_cfg.name = "ESP32 Audio Device";
  a2dp_input.begin(a2dp_cfg);
#else
  // register callback
  a2dp_in_sink.set_stream_reader(read_data_stream, false);
  a2dp_in_sink.set_volume(0.3);
  // Start Bluetooth Audio Receiver
  a2dp_in_sink.set_auto_reconnect(false);
  a2dp_in_sink.start("ESP32 Audio Device");
#endif /* USE_A2DP_STREAM */


#if USE_A2DP_STREAM
  mixer.add(a2dp_input);
#else
  // mixer.add(a2dp_in_sink);
#endif /* USE_A2DP_STREAM */
  mixer.add(i2s_input);
  mixer.begin(info);
}

```

**Reply by Mukhin22** [reply]
But when I'm trying to run it I have the following errors:
```
"17:42:17.977 -> Starting A2DP...
17:42:17.977 -> Starting I2S output...
17:42:17.977 -> Starting I2S input...
17:42:59.468 -> Mixing audio streams...
17:42:59.468 -> E (41524) I2S: i2s_read(2264): RX mode is not enabled
17:42:59.500 -> [E] I2SESP32.h : 100 - size_t audio_tools::I2SDriverESP32::readBytes(void*, size_t)"
```

**Reply by pschatzmann** [reply]
You obviously did not open it in RX mode!
Try the following example first to understand how this is working:
https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-stream/streams-a2dp-serial/streams-a2dp-serial.ino

**Reply by Mukhin22** [reply]
Sorry, it was my mistake here, wrong copy direction:
`StreamCopy copier(mixer, i2s_output);`

**Comment by Mukhin22** [comment]
Hello, Phil, have to come back to you again. I have the same configuration: I'm using ESP32-DevKitC(controller), MAX98357A(amplifier), INMP441(micro).
But still cannot find a way to get a clean sound on output i2s with such a code:
```
#include "AudioTools.h"
#include "AudioLibs/AudioA2DP.h"

//#pragma once

//#define I2S_BUFFER_COUNT 8
//#define I2S_BUFFER_SIZE 256

// config params
#define DEBUG_EN 1

AudioBaseInfo info_speaker;
AudioBaseInfo info_micro;

const uint16_t sample_rate = 44100;
const uint16_t channels = 2;
const uint16_t bits_per_sample = 16;
const uint16_t micro_bits_per_sample = 32; // Note: 16 bits don't work also. INMP441 micro has 24/32 bits representation on i2s

I2SStream i2s_input;
I2SStream i2s_output;
InputMixer<int16_t> mixer;
A2DPStream a2dp_input;
ConverterFillLeftAndRight<int16_t> filler(audio_tools::Auto); // fill both channels
                                        
//FormatConverterStream conv(i2s_input);
StreamCopy copier(i2s_output, mixer);

void setup(void) {  
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  info_speaker.channels = channels;
  info_speaker.sample_rate = sample_rate;
  info_speaker.bits_per_sample = bits_per_sample;

  info_micro.channels = channels;
  info_micro.sample_rate = sample_rate;
  info_micro.bits_per_sample = micro_bits_per_sample;
  
#if DEBUG_EN
  Serial.println("Starting A2DP...");
#endif

//  volume.begin(volume.defaultConfig());
//  volume.setVolume(1.0);
  a2dp_input.setNotifyAudioChange(i2s_output);

#if DEBUG_EN
  Serial.println("Starting I2S input...");
#endif
  // Configure and start the input i2s
  auto i2s_input_config = i2s_input.defaultConfig(RX_MODE);
  i2s_input_config.copyFrom(info_micro);
  i2s_input_config.i2s_format = I2S_STD_FORMAT;
  i2s_input_config.is_master = true;
  i2s_input_config.port_no = 0;
  i2s_input_config.pin_bck = 15;
  i2s_input_config.pin_ws = 14;
  i2s_input_config.pin_data = 22;
  i2s_input.begin(i2s_input_config);

#if DEBUG_EN
  Serial.println("Starting I2S output...");
#endif
  // Configure and start the output i2s
  auto i2s_output_config = i2s_output.defaultConfig(TX_MODE);
  i2s_output_config.copyFrom(info_speaker);
  i2s_output_config.i2s_format = I2S_STD_FORMAT;
  i2s_output_config.is_master = true;
  i2s_output_config.port_no = 1;
  i2s_output_config.pin_bck = 18;
  i2s_output_config.pin_ws = 19;
  i2s_output_config.pin_data = 23;
  i2s_output.begin(i2s_output_config);
  
  // begin converter stream from 32 bit in to 16 bit out 
  // conv.begin(info_micro, info_speaker);
  // setup the a2dp input audio stream
  auto a2dp_cfg = a2dp_input.defaultConfig(RX_MODE);
  a2dp_cfg.name = "ESP32 Hearing Aid Speaker";
  a2dp_input.begin(a2dp_cfg);
  a2dp_input.setVolume(0.2);

#if DEBUG_EN
  Serial.println("Mixing audio streams...");
#endif
  // Add streams to mixing of inputs  and start 
  mixer.add(i2s_input);
  mixer.add(a2dp_input);
  mixer.begin(i2s_output_config);
}

void loop() {
  copier.copy();
}
```
I assume the reason is I can not correctly convert INMP441 i2s data format.
I was trying to use FormatConverterStream and ConverterFillLeftAndRight according to the lib documentation and examples you have. But still can't find a way to get the clear sound on output. (connection of the hardware is ok, because with same hardware pinout with adapted i2s0 to i2s1 example it works perfectly. 
I'm hearing both the sound from micro and a2dp, but they noisy as hell.
Possibly the way I used FormatConverterStream and ConverterFillLeftAndRight is wrong, or the way I use inputMixer. Is it possible for you to look with expert eye?

**Reply by Mukhin22** [reply]
As for me such configuration should work correct according to "Converting input side example", but it has no sound at all and no error messages:
```

I2SStream i2s_input;
I2SStream i2s_output;
InputMixer<int16_t> mixer;
FormatConverterStream conv(i2s_input);
StreamCopy copier(i2s_output, mixer);
 conv.begin(info_micro, info_speaker);
  mixer.add(conv);
  mixer.add(a2dp_input);
  mixer.begin(i2s_output_config);

void loop() {
  copier.copy();
}

```

**Reply by pschatzmann** [reply]
Go back one step and try to make sure that you get a clean 16 bit stereo signal from the INMP441!
Only then you can go to the next step and implement the mixing code...
If I got your logic right you read the INMP441 with 32 bits, so you need to convert it to 16 bits before giving this stream as input...

**Reply by Mukhin22** [reply]
The reason I'm trying to read it with 32 bits, Is that I'm not getting clear sound with 16 bits when I'm trying to use simply InputMixer to mix i2s input stream and a2dp input stream.

I confirmed that that I'm getting clean 16 bit stereo signal with this example firstly, and currently this also works:
https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-stream/streams-i2s-i2s-2/streams-i2s-i2s-2.ino

**Reply by pschatzmann** [reply]
Please note that you if you add 2 inputs to the mixer, it will only provide any data if both inputs actually provide data. 
Did you actualy open a bluetooth connection e.g. with your mobile phone.
In the second last example you still try to mix 32 bit data with 16 bit data. This is not supported and you need to change the format. eg. with my updated FormatConverterStream 

see https://www.pschatzmann.ch/home/2023/04/01/arduino-audiotools-formatconverterstream/

**Reply by Mukhin22** [reply]
Yes, mixer is working, i can hear the data from both sources, but  the out stream interrupted and noisy.
I'll check what you sent, thanks

**Reply by Mukhin22** [reply]
It work in that way when I'm setting up the inmp441 microphone to 16 bit.
I was also trying to play with buffer sizes of 2s and a2dp and arrays count, it changes the noise, but still not clear.

**Comment by pschatzmann** [comment]
Just look at my link: it shows haw you can read 32 bits from I2S and convert it to the needed 16 bits e.g. for the mixer

**Comment by Mukhin22** [comment]
Its working (i hear the sound from both sources and it's clean) for about 5 seconds after bluetooth connection, when I'm using format converter as you described.  It sends me such message in logs and esp32 restarted after that:
assert failed: host_recv_pkt_cb hci_hal_h4.c:548 (0)
with following backtrace, esp32 exception decoder fails to analyse this back trace:
Backtrace: 0x40083981:0x3ffdc940 0x40092eb1:0x3ffdc960 0x40099715:0x3ffdc980 0x4011b136:0x3ffdcab0 0x401752fa:0x3ffdcad0 0x40164802:0x3ffdcaf0 0x4001791d:0x3ffdcb10 0x40165085:0x3ffdcb30 0x40165162:0x3ffdcb70 0x400188f5:0x3ffdcba0 0x40164a47:0x3ffdcbe0 0x4016843d:0x3ffdcc00 0x400869d4:0x3ffdcc20 0x40019d11:0x3ffdcc50 0x40055b4d:0x3ffdcc70 0x40162f83:0x3ffdcc90 0x401635e9:0x3ffdccb0

All I was manged to find about this assert is this suggests that there is an issue with the Bluetooth packet reception function. May it be the issue of heap over or stack overflow or may be some software problem?

This is my current code state:

```
#include "AudioTools.h"
#include "AudioLibs/AudioA2DP.h"
#include "AudioConfigLocal.h"
//#include "AudioLibs/AudioKit.h"

// TODO: try to start on single input
// config params
#define DEBUG_EN 1

AudioBaseInfo info_speaker;
AudioBaseInfo info_micro;

const uint16_t sample_rate = 44100;
const uint16_t channels = 2;
const uint16_t bits_per_sample = 16;
const uint16_t micro_bits_per_sample = 32; // INMP441 micro has 24/32 bits representation on i2s

I2SStream i2s_input;
I2SStream i2s_output;
InputMixer<int16_t> mixer;
A2DPStream a2dp_input;
FormatConverterStream converter(i2s_input);                                        
StreamCopy copier(i2s_output, mixer);

void setup(void) {  
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Error);

  info_speaker.channels = channels;
  info_speaker.sample_rate = sample_rate;
  info_speaker.bits_per_sample = bits_per_sample;

  info_micro.channels = channels;
  info_micro.sample_rate = sample_rate;
  info_micro.bits_per_sample = micro_bits_per_sample;
  
#if DEBUG_EN
  Serial.println("Starting A2DP...");
#endif

#if DEBUG_EN
  Serial.println("Starting I2S input...");
#endif
  // Configure and start the input i2s
  auto i2s_input_config = i2s_input.defaultConfig(RX_MODE);
  i2s_input_config.copyFrom(info_micro);
  i2s_input_config.i2s_format = I2S_STD_FORMAT;
  i2s_input_config.is_master = true;
  i2s_input_config.port_no = 0;
  i2s_input_config.pin_bck = 14;
  i2s_input_config.pin_ws = 15;
  i2s_input_config.pin_data = 22;
  i2s_input.begin(i2s_input_config);

  converter.begin(info_micro, info_speaker);
#if DEBUG_EN
  Serial.println("Starting I2S output...");
#endif
  // Configure and start the output i2s
  auto i2s_output_config = i2s_output.defaultConfig(TX_MODE);
  i2s_output_config.copyFrom(info_speaker);
  i2s_output_config.i2s_format = I2S_STD_FORMAT;
  i2s_output_config.is_master = true;
  i2s_output_config.port_no = 1;
  i2s_output_config.pin_bck = 18;
  i2s_output_config.pin_ws = 19;
  i2s_output_config.pin_data = 23;
  i2s_output.begin(i2s_output_config);
  // setup the a2dp input audio stream
  auto a2dp_cfg = a2dp_input.defaultConfig(RX_MODE);
  a2dp_cfg.name = "ESP32 Hearing Aid Speaker";
  a2dp_input.begin(a2dp_cfg);
  //a2dp_input.setVolume(0.1);

#if DEBUG_EN
  Serial.println("Mixing audio streams...");
#endif
  // Add streams to mixing of inputs  and start 
  mixer.add(a2dp_input);
  mixer.add(converter);
  mixer.begin(info_speaker);
}

void loop() {
  copier.copy();
}
```

**Reply by pschatzmann** [reply]
Hmm, difficult to say what's causing this and not having a decoded stack-trace does not help.
The only thing that you could try is to decrease the copy size e.g. with StreamCopy copier(i2s_output, mixer, 256); and maybe add some log statements in the loop that show the available heap.

Maybe increasing the i2s input buffer size might also be worth a trial...

What's happening if you commend out the  mixer.add(converter); and you only get the a2dp input thru the mixer ?

**Comment by Mukhin22** [comment]
Yes, the heap contentiously decreasing from 50000 available to 30000 and at the point of 30k it gives this "assert failed: host_recv_pkt_cb hci_hal_h4.c:548 (0)". Even when commend out the mixer.add(converter);

**Reply by Mukhin22** [reply]
StreamCopy copier(i2s_output, mixer, 256); such thing only gives bad audio quality, but the issue behavior is the same.

**Reply by Mukhin22** [reply]
And increasing the I2S_BUFFER_SIZE from 512 to 2048 leads to longer working time, but this issue still triggered after this "longer time".

**Comment by pschatzmann** [comment]
Honestly I did not expect this outcome and I would not know how to solve this.
My gut feeling is that the root cause might be, that the processing is distributed over the 2 cores.

Maybe using the OutputMixer should avoid this issue, so it might be worth a trial: The starting point would be [this sketch](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-basic-api/basic-a2dp-i2s/basic-a2dp-i2s.ino) and you would call the mixer output logic in the callback replacing the i2s output.

**Reply by Mukhin22** [reply]
Thank you for the help.
I saw couple of threads about similar issues in other lib, but it's a little bit hard to dive deeper into the lib internals in arduino ide, maybe they will be interesting to you:
https://github.com/bluekitchen/btstack/issues/357
https://github.com/espressif/esp-idf/issues/1322
I'll try the output mixer.

**Reply by pschatzmann** [reply]
Yes, I have seen this: just confirm that in your test you left the log level at AudioLogger::Error

**Reply by pschatzmann** [reply]
When using the output mixer, avoid to use the A2DPStream because this is using a FreeRTOS queue: I guess this might cause the issue...

**Reply by Mukhin22** [reply]
> Yes, I have seen this: just confirm that in your test you left the log level at AudioLogger::Error

Yes, I haven't changed it.

**Comment by pschatzmann** [comment, marked as answer]
I tried to set up a POC on the AudioKit. Initially, I was stumbling a bit over the big array size that we need to process.
The following sketch works like a charm:

```C++

#include "BluetoothA2DPSink.h"
#include "AudioTools.h"
#include "AudioLibs/AudioKit.h"

AudioInfo info(44100, 2, 16);
BluetoothA2DPSink a2dp_sink;
AudioKitStream i2s;
OutputMixer<int16_t> mixer(i2s, 2);
const int buffer_size = 1024;  
uint8_t buffer[buffer_size];

// Write data to mixer in callback
void read_data_stream(const uint8_t *data, uint32_t length) {
  int open = length;
  int pos = 0;
  // limit write size
  while(open>0){
    int write_size = min(open, 1024);
    int bytes_read = i2s.readBytes(buffer, write_size);
    // mix data
    mixer.write(data+pos, write_size);
    mixer.write(buffer, write_size);
    mixer.flushMixer();
    open -= write_size;
    pos += write_size;
  }
}

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);

  // register callback
  a2dp_sink.set_stream_reader(read_data_stream, false);

  // setup input and output
  auto cfg = i2s.defaultConfig(RXTX_MODE);
  cfg.copyFrom(info);
  cfg.input_device = AUDIO_HAL_ADC_INPUT_LINE2;
  i2s.begin(cfg);

  // setup Output mixer in RAM
  mixer.begin(buffer_size, RAM);

  // Start Bluetooth Audio Receiver
  a2dp_sink.set_auto_reconnect(false);
  a2dp_sink.start("a2dp-i2s");

}

void loop() { delay(1000); }
```
Just replace the AudioKitStream with your I2SStreams and you  need to deal with the 32 bits of the microphone. Something that was not necessary for my hardware...

**Reply by Mukhin22** [reply]
Hi, Phil, many thanks to you for such a detailed example. It really works like a charm after adaptations I need for my hardware, has a small delay, but I don't think its a big problem to fix.

**Comment by Mukhin22** [comment]
Hi, Phil, thank you for this magnificent lib again.
I've the situation when micro is very close to output speaker. I was searching but not found here some echo cancellation methods in lib or examples. 
Can you recommend something for echo cancellation?

**Comment by pschatzmann** [comment]
This has nothing to do with echo cancellation: I suggest that you google "prevent audio feed back"!
